### PR TITLE
Particle/mpm fluid

### PR DESCRIPTION
**Pull request by alecontri**
**📝 Description**
In this branch I added the constitutive laws describing the behaviour of both a weakly compressible and an incompressible (using the existing UP element) newtonian fluid to the ParticleMechanicsApplication. Moreover I modified te UP element in order to describe a general total incompressible material [Holzapfel,ch.8] extending the previous version of the stabilization, that only allowed materials that had a Young Modulus and a Poisson Ratio, to the new material. Minor corrections were added, too.

Please mark the PR with appropriate tags: 
- Dispacement-based Newtonian MPM material

**🆕 Changelog**
- Addition of a MPM law for newtonian fluids 
- Modification of the UP element to totally incompressible (with correction of the Damping and Mass matrix computation)
- Extension of stabilization parameter to shallow water simulations
- Addition of a python script process for a peculiar initial velocity distribution (to be deleted later)
- Addition of the missing lines of code for the reactions computations

### CODE REVIEW

File: applications/ParticleMechanicsApplication/particle_mechanics_application_variables.h [hunk 1]
```diff
@@ -19,6 +19,7 @@
 #include "includes/define.h"
 #include "includes/constitutive_law.h"
 #include "includes/variables.h"
+#include "includes/cfd_variables.h"
```
Comment (philbucher) [inline applications/ParticleMechanicsApplication/particle_mechanics_application_variables.h]:
```suggestion
```

please do not include this globally, it increases the compilation time
=> it should only be included when you actually need it

Comment (alecontri) [reply applications/ParticleMechanicsApplication/particle_mechanics_application_variables.h]: Thanks @philbucher. I use it in the constitutive law and the UP element. Adding it just there is ok? Is it better in the .hpp or the .cpp?

Comment (philbucher) [reply applications/ParticleMechanicsApplication/particle_mechanics_application_variables.h]: Yes only in the files where you need it. if possible only in the cpp. If you also need it in the hpp then only there (since the hpp gets included in the cpp)

Comment (alecontri) [reply applications/ParticleMechanicsApplication/particle_mechanics_application_variables.h:22]: Thank you. Sorry, but now I see that not all modifications I made were reported. Is it ok if I close the pull request and re-open when it will be right? Sorry, my fault, but I am new to this

**Review by lauramoremar — APPROVED** [review]

**Comment by VeronikaSinger** [comment]
Hi @alecontri 
nice feature for the MPM application 👍 
Do you have any validation test, which could be added to the example repository, too?

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.cpp [hunk 2]
```diff
@@ -1255,7 +1255,14 @@ void UpdatedLagrangian::CalculateDampingMatrix( MatrixType& rDampingMatrix, cons
...
     //resizing as needed the LHS
-    unsigned int matrix_size = number_of_nodes * dimension;
+    unsigned int matrix_size;
+    if (rCurrentProcessInfo["pressure_dofs"]) {
+        matrix_size = number_of_nodes * (dimension + 1);
+    }
+    else {
+        matrix_size = number_of_nodes * dimension;
+    }
+
```
Comment (VeronikaSinger) [inline applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.cpp:1265]: Why is this change needed here?

File: applications/ParticleMechanicsApplication/python_scripts/assign_initial_velocity_to_particle_process2.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        self.model_part = Model[settings["model_part_name"].GetString()]
+        self.modulus = settings["modulus"].GetDouble()
+        self.velocity_direction = settings["direction"].GetVector()
+        self.velocity = self.modulus * self.velocity_direction
+        gravity = 9.81
+        self.height = 1.0
+        frequency = 0.89
+        self.wave_number = math.pi
+        self.constant = gravity * self.modulus / ( 2.0 * frequency * math.cosh( self.wave_number * self.height ) ) * self.wave_number
+
```
Comment (VeronikaSinger) [inline applications/ParticleMechanicsApplication/python_scripts/assign_initial_velocity_to_particle_process2.py:54]: I guess you can define this in the default parameters. 
Isn't it possible to reuse the assign_initial_velocity_to_particle_process and just add the the additional variables in a derived process.

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_UP.cpp [hunk 4]
```diff
@@ -415,9 +416,9 @@ void UpdatedLagrangianUP::CalculateAndAddRHS(
     const double& rIntegrationWeight,
     const ProcessInfo& rCurrentProcessInfo)
 {
-    rVariables.detF0   *= rVariables.detF;
-    double determinant_F = rVariables.detF;
-    rVariables.detF = 1; //in order to simplify updated and spatial lagrangian
+    //rVariables.detF0   *= rVariables.detF;
+    //double determinant_F = rVariables.detF;
+    //rVariables.detF = 1; //in order to simplify updated and spatial lagrangian
```
Comment (VeronikaSinger) [inline applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_UP.cpp:422]: why is the update not needed here any more?

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_UP.cpp [hunk 5]
```diff
@@ -536,33 +505,10 @@ void UpdatedLagrangianUP::CalculateAndAddPressureForces(VectorType& rRightHandSi
...
-        for ( unsigned int j = 0; j < number_of_nodes; j++ )
-        {
-            const double& pressure = r_geometry[j].FastGetSolutionStepValue(PRESSURE);
-
-            // TODO: Check what is the meaning of this equation
-            rRightHandSideVector[index_p] += (1.0/(delta_coefficient * bulk_modulus)) * r_N(0, i) * r_N(0, j) * pressure * rIntegrationWeight / (rVariables.detF0/rVariables.detF) ; //2D-3D
-        }
 
-        rRightHandSideVector[index_p] -=  coefficient/delta_coefficient * r_N(0, i) * rIntegrationWeight / (rVariables.detF0/rVariables.detF);
+        rRightHandSideVector[index_p] -= (1.0 - 1.0 / rVariables.detFT) * r_N(0, i) * rIntegrationWeight;
```
Comment (VeronikaSinger) [inline applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_UP.cpp:512]: can you explain this in more detail?

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_UP.cpp [hunk 6]
```diff
@@ -1037,7 +1017,7 @@ void UpdatedLagrangianUP::CalculateMassMatrix( MatrixType& rMassMatrix, const Pr
         for (IndexType i = 0; i < number_of_nodes; ++i) {
             for (IndexType k = 0; k < dimension; ++k)
             {
-                const IndexType index = i * dimension + k;
+                const IndexType index = i * (dimension + 1) + k;
                 rMassMatrix(index, index) = N[i] * mMP.mass;
             }
         }
```
Comment (VeronikaSinger) [inline applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_UP.cpp:1023]: It's really hard to follow all the changes in the UP-Element. We need to be sure that it still works - did you run further validaton examples?

**Review by VeronikaSinger — CHANGES_REQUESTED** [review]

**Review by VeronikaSinger — APPROVED** [review]

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.cpp (hunk 2 above)
Comment (alecontri) [reply applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.cpp:1265]: Even if the merging was not in the master, I reply as promised (sorry for being late). The change is due to the fact that in the master version of the UP element a degree of freedom is added (pressure) but the mass and damping matrix dimensions were not updated by consequence causing errors in the dynamic simulations.

File: applications/ParticleMechanicsApplication/python_scripts/assign_initial_velocity_to_particle_process2.py (hunk 3 above)
Comment (alecontri) [reply applications/ParticleMechanicsApplication/python_scripts/assign_initial_velocity_to_particle_process2.py:54]: At that time, I cannot find a way to introduce such a complex law as initial velocity using default processes. However, this process was only temporary needed (assign_initial_velocity_to_particle_process2.py) and the change can be discarded with no problem (deleting assign_initial_velocity_to_particle_process2.py).

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_UP.cpp (hunk 4 above)
Comment (alecontri) [reply applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_UP.cpp:422]: In the version I studied, this change was not required at all. Moreover, given the change done here and the values defined in the classes CalculatePUCoefficient() and CalculatePUDeltaCoefficient() the resulting formulation is the same as the modified, but seemed to me more elegant since resulting directly from the discretization of the mixed formulation presented in Nonlinear Solid Mechanics: An Continuum approach for Engineer, Chapter 8.

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_UP.cpp (hunk 6 above)
Comment (alecontri) [reply applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_UP.cpp:1023]: I see. Here it was just a problem deriving from the fact that a degree of freedom is added in the UP element but the Mass Matrix dimension is not updated accordingly. For the complete formulation, I attach my thesis hoping it will describe better any change adopted
[tesi_alessandro_contri.pdf](https://github.com/KratosMultiphysics/Kratos/files/7942023/tesi_alessandro_contri.pdf)

Comment (loumalouomega) [reply applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_UP.cpp:1023]: Nice document
